core/src/main/java/io/questdb/griffin/SqlParser.java

```diff
@@ -882,24 +879,37 @@ private ExecutionModel parseCreateMatView(
             final int queryStartPos = lexer.getPosition();
 
             // parse mat view query
+            tok = tok(lexer, "'with' or 'select'");
+            if (isWithKeyword(tok)) {
+                parseWithClauses(lexer, topLevelWithModel, sqlParserCallback, null);
+                // CTEs require SELECT to be specified
+                expectTok(lexer, "select");
+            }
+            lexer.unparseLast();
             final QueryModel qm = parseDml(lexer, null, lexer.getPosition(), true, sqlParserCallback, null);
             final QueryModel nestedModel = qm.getNestedModel();
-            if (nestedModel.getSampleByFrom() != null) {
-                throw SqlException.position(lexer.getPosition()).put("FROM is not supported for materialized views");
-            }
-            if (nestedModel.getSampleByTo() != null) {
-                throw SqlException.position(lexer.getPosition()).put("TO is not supported for materialized views");
-            }
 
-            if (nestedModel.getSampleByFill() != null && nestedModel.getSampleByFill().size() > 0) {
-                throw SqlException.position(lexer.getPosition()).put("FILL is not supported for materialized views");
+            // check for all nested models
+            QueryModel m = nestedModel;
+            while (m != null) {
+                if (m.getSampleByFrom() != null || m.getSampleByTo() != null) {
+                    final int position = m.getSampleByFrom() != null ? m.getSampleByFrom().position : m.getSampleByTo().position;
+                    throw SqlException.position(position).put("FROM-TO is not supported for materialized views");
+                }
+                final ObjList<ExpressionNode> sampleByFill = m.getSampleByFill();
+                if (sampleByFill != null && sampleByFill.size() > 0) {
+                    throw SqlException.position(sampleByFill.get(0).position).put("FILL is not supported for materialized views");
+                }
+                m = m.getNestedModel();
```

Comment:
do we support union or any other set operation ?

Reply:
there can be a fill in a union model, which is unchecked. Typical through model validation has to:
- traverse "nestedModel"
- traverse "unionModel"
- traverse "joinModels" from index 1 onwards

Reply:
anyway, i think this area of code is most questionable for other reasons

Reply:
> there can be a fill in a union model, which is unchecked

Yes, we discussed that. In principle, unions in the query or left-hand of the join are problematic as they need to be surrounded with a TIMESTAMP clause and likely to break incremental refreshes. On the other hand, unions in right-hand of the joins aren't a problem unless the base table is used there.